### ISSUE TITLE
Brittle logic in renaming dict

### ISSUE THREAD

**Issue by jbusecke** [jbusecke/cmip6_preprocessing#52]
As nicely pointed out in #51, dimension names that could also appear as coordinate names (e.g. lon/lat are used in both contexts), need to be at the end of the renaming dicts entry for e.g. `x`. Entering new names could lead to unintended behavior. 

Just pinning this here as a reminder for the next refactor.

**Comment by jbusecke** [comment]
Ok I am convinced that I should refactor the renaming function. 

A brief motivation:

I recently found this dataset
```python
from cmip6_preprocessing.utils import google_cmip_col
col = google_cmip_col()
cat_test = col().search(
    table_id='Ofx',
    grid_label='gn',
    source_id='GFDL-CM4',
    experiment_id='historical', # Removing this results in a LOT of datasets, which takes forever. Remove this at the end to get the maximum amount of matches
    variable_id= 'areacello',
)
ddict_test = cat_test.to_dataset_dict()
_, ds = ddict_test.popitem()
ds
```
<img width="623" alt="image" src="https://user-images.githubusercontent.com/14314623/115412396-43505e80-a1c2-11eb-8ddf-7fd7f5c7c1b3.png">

The current renaming fails
```python
from cmip6_preprocessing.preprocessing import rename_cmip6
rename_cmip6(ds)
```
<img width="617" alt="image" src="https://user-images.githubusercontent.com/14314623/115412656-84487300-a1c2-11eb-93f4-b9c99140b167.png">

I think the problem here is that the [current logic](https://github.com/jbusecke/cmip6_preprocessing/blob/cf59354e9c856e043c2ead66e27ad632b9aaa60a/cmip6_preprocessing/preprocessing.py#L78) stops as soon as any `x` or `y` are present in the whole dataset. 

A quick an dirty fix could be to instead apply the renaming on all variables, but I wanted to check if there is a more elegant way, we can implement here. 

@dcherian, how would you envision cf-xarray to fit in here? I did not entirely follow what you said in https://github.com/xarray-contrib/cf-xarray/issues/133, but if I spend time on this (which I have to kinda), I would like to get it right this time.

**Comment by dcherian** [comment]
What is the expected output here? Rename `xh`, `yh` to `x`, `y`? I'll have to see attributes and I can't get it to run on the pangeo cloud

![image](https://user-images.githubusercontent.com/2448579/115417115-95d74e00-a1b5-11eb-952f-e9fa561f0854.png)

Can you upload the notebook to a gist

**Comment by jbusecke** [comment]
I have been trying to play around with the [renaming feature]() but am not sure what the best way to deal with this is. 
Would I create a 'dummy datasets' that has all cf attributes and then use `ds.cf.rename_like(dummy_dataset)`?

**Comment by jbusecke** [comment]
> What is the expected output here? Rename xh, yh to x, y? 

Yep exactly.

> I'll have to see attributes and I can't get it to run on the pangeo cloud

Ah sorry, that was my bad. I edited the code above, could you try again with:
```python
col = intake.open_esm_datastore("https://storage.googleapis.com/cmip6/pangeo-cmip6.json")
```

**Comment by jbusecke** [comment]
If that is of help the output of `ds.cf` is:
```
Coordinates:
- CF Axes:   X: ['lon', 'lon_bnds', 'xh']
             Y: ['lat', 'lat_bnds', 'yh']
             Z, T: n/a

- CF Coordinates:   longitude: ['lon', 'lon_bnds', 'xh']
                    latitude: ['lat', 'lat_bnds', 'yh']
                    vertical, time: n/a

- Cell Measures:   area, volume: n/a

- Standard Names:   latitude: ['lat']
                    longitude: ['lon']

Data Variables:
- Cell Measures:   area, volume: n/a

- Standard Names:   cell_area: ['areacello']
```

**Comment by dcherian** [comment]
``` python
rename_dict = {
        # dim labels (order represents the priority when checking for the dim labels)
        "x": ["x", "i", "ni", "xh", "nlon", "lon", "longitude"],
        "y": ["y", "j", "nj", "yh", "nlat", "lat", "latitude"],
        "lev": ["lev", "deptht", "olevel", "zlev", "olev", "depth"],
        "bnds": ["bnds", "axis_nbounds", "d2"],
        "vertex": ["vertex", "nvertex", "vertices"],
        # coordinate labels
        "lon": ["lon", "longitude", "nav_lon"],
        "lat": ["lat", "latitude", "nav_lat"],
        "lev_bounds": [
            "lev_bounds",
            "deptht_bounds",
            "lev_bnds",
            "olevel_bounds",
            "zlev_bnds",
        ],
        "lon_bounds": [
            "bounds_lon",
            "bounds_nav_lon",
            "lon_bnds",
            "x_bnds",
            "vertices_longitude",
        ],
        "lat_bounds": [
            "bounds_lat",
            "bounds_nav_lat",
            "lat_bnds",
            "y_bnds",
            "vertices_latitude",
        ],
        "time_bounds": ["time_bounds", "time_bnds"],
    }

cf_rename_dict = {
"X": "x",
"Y": "y",
"Z": "lev",
"vertical": "lev",
"longitude": "lon",
"latitude": "lat",
}
```

In the end you want
``` python
ds.cf.rename(cf_rename_dict)
```

(note I don't think bounds are supported yet but we can add that).

will work as long as `cf_xarray` can identify the variable on the RHS of `rename_dict` as being tagged with appropriate attributes. For e.g. `"x", "i"` etc. should have `axis: "X"`. 

There is also `ds.cf.guess_coord_axis(verbose=True)` which will try to autoguess and add attributes. That could make your life easier and is probably what you want once we add all the heuristics you need for cmip output.

Basically we would need to teach `cf_xaray` heuristics to identify `rename_dict` RHS variable names as the LHS of `cf_rename_dict`

cc @malmans2 who's worked on renaming the most. Mattia, the bounds renaming is an interesting use-case.

**Comment by dcherian** [comment]
@jbusecke does `ds.cf.rename({"X": "x", "Y": "y"})` help?

**Comment by dcherian** [comment]
Yeah this won't work. There are multiple variables tagged as `axis: 'X'` (['lon', 'lon_bnds', 'xh']) so `.cf.rename` won't work. :/ 

I guess step one is to remove the `axis` attribute from `longitude`-like variables. Then attempt renaming with cf_xarray.

**Comment by jbusecke** [comment]
Thanks for the detailed tips. Ill try that in a bit and report back!

**Comment by dcherian** [comment]
Haha this dataset is really something.

`xh, lon, lon_bounds` are all identifies as both `'X'` and `'longitude'` which is technically right but practically useless when it comes to automated renaming :P.

`xh` has these attributes (cf_xarray parses units to identify it as longitude)
```
cartesian_axis :
    X
long_name :
    h point nominal longitude
units :
    degrees_east
```

`lon` has these attributes
```
axis :
    X
bounds :
    lon_bnds
cell_methods :
    y: point x: point
long_name :
    longitude
standard_name :
    longitude
units :
    degrees_east
```

We definitely want to remove `axis: X` from `lon` for things to work.

Alternate (and better) solution is to add an sgrid attribute and parse that.

**Comment by jbusecke** [comment]
> Haha this dataset is really something.

This should really be the `cmip6_preprocessing` tagline!

**Comment by dcherian** [comment]
we can have a quick video call to chat about this if you want.

**Comment by jbusecke** [comment]
I am about to wrap up over here, and tomorrow is quite busy on the family side. Any chance we can pick this up on Thu?

**Comment by dcherian** [comment]
How about a half hour before/after the sgrid meeting on Friday.

**Comment by jbusecke** [comment]
I decided against using a renaming logic based on cf_xarray for now. My idea is to instead rename manually and then parse cf attributes (which is not properly done in all existing datasets. Ill close this now and add a new issue.
